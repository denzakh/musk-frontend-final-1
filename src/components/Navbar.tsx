import { NavLink, Link } from 'react-router-dom';
import favoriteImg from '../assets/favorite.svg';
import burgerImg from '../assets/burger.svg';
import { useState } from 'react';

const categories = [
    'general',
    'sports',
    'technology',
    'business',
    'health',
    'entertainment',
];

const Navbar = () => {
    const [open, setOpen] = useState<boolean>(false);

    return (
        <nav className='relative bg-white dark:bg-gray-800 shadow p-4 flex justify-between items-center gap-4'>
            <Link to={'/'}>
                <div className='flex border-blue-500 border-2 font-semibold'>
                    <div className='bg-blue-500 text-white px-2 py-1'>Musk</div>
                    <div className='text-blue-500 bg-white px-2 py-1'>News</div>
                </div>
            </Link>
            <div className={open ? 'block md:block' : 'hidden md:block'}>
                <div className='bg-white flex gap-8 md:gap-4 text-2xl md:text-base absolute md:relative flex-col md:flex-row top-18 md:top-0 right-1 md-right-0 z-10 p-8 md:p-0 shadow-xl md:shadow-none rounded-xl'>
                    {categories.map((cat) => (
                        <NavLink
                            key={cat}
                            to={`/category/${cat}`}
                            className={({ isActive }) =>
                                isActive
                                    ? 'text-blue-500 font-semibold'
                                    : 'text-gray-600 dark:text-gray-300'
                            }
                            onClick={() => setOpen(false)}
                        >
                            {cat.charAt(0).toUpperCase() + cat.slice(1)}
                        </NavLink>
                    ))}
                </div>
            </div>

            <div className='ml-[auto] md:ml-0'>
                <NavLink
                    to='/favorites'
                    className={({ isActive }) =>
                        isActive
                            ? 'text-blue-500 font-semibold'
                            : 'text-gray-600 dark:text-gray-300'
                    }
                >
                    <div className='flex items-center gap-2'>
                        <img
                            src={favoriteImg}
                            alt={''}
                            width={30}
                            height={30}
                            className=''
                        />
                        <span className='hidden md:inline'>Favorites</span>
                    </div>
                </NavLink>
            </div>
            <div className='flex items-center md:hidden pl-2'>
                <button className='' onClick={() => setOpen((open) => !open)}>
                    <img
                        src={burgerImg}
                        alt={''}
                        width={30}
                        height={30}
                        className=''
                    />
                </button>
            </div>
        </nav>
    );
};

export default Navbar;
