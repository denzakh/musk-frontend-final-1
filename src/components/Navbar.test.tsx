import { render, screen, fireEvent } from '@testing-library/react';
import { vi } from 'vitest';
import Navbar from './Navbar';
import { BrowserRouter } from 'react-router-dom';
import { ReactNode } from 'react';

vi.mock('../assets/favorite.svg', () => ({ default: 'mocked-favorite.svg' }));
vi.mock('../assets/burger.svg', () => ({ default: 'mocked-burger.svg' }));
vi.mock('../assets/search.svg', () => ({ default: 'mocked-search.svg' }));
vi.mock('../assets/sun.svg', () => ({ default: 'mocked-sun.svg' }));
vi.mock('../assets/moon.svg', () => ({ default: 'mocked-moon.svg' }));

export const mockToggle = vi.fn();

// Creamos un mock para useNavigate y lo exportamos para verificarlo luego
const mockNavigate = vi.fn();

// Mock de react-router-dom
vi.mock('react-router-dom', async () => {
    const actual = await vi.importActual<typeof import('react-router-dom')>(
        'react-router-dom'
    );

    return {
        ...actual,
        useParams: vi.fn(() => ({ category: 'general' })),
        useNavigate: () => mockNavigate,
        NavLink: vi.fn(
            ({
                to,
                children,
                className,
                onClick,
            }: {
                to: string;
                children: ReactNode;
                className:
                    | string
                    | (({ isActive }: { isActive: boolean }) => undefined);
                onClick: () => void;
            }) => (
                <a
                    href={to}
                    className={
                        typeof className === 'function'
                            ? className({ isActive: false })
                            : className
                    }
                    data-testid={`link-${to}`}
                    onClick={(e) => {
                        e.preventDefault(); // Prevenir navegación real
                        if (onClick) onClick(); // Simular click handler
                        mockNavigate(to); // Simular navegación
                    }}
                >
                    {children}
                </a>
            )
        ),
    };
});

beforeEach(() => {
    vi.clearAllMocks();
    vi.resetModules();
});

describe('Navbar', async () => {
    it('snapshot', async () => {
        vi.doMock('../hooks/useDarkMode', () => {
            return {
                useDarkMode: vi.fn(() => [false, mockToggle]),
            };
        });
        const { default: Navbar } = await import('./Navbar');

        const { asFragment } = render(
            <BrowserRouter>
                <Navbar />
            </BrowserRouter>
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it('existe icon dark en el theme dark', async () => {
        vi.doMock('../hooks/useDarkMode', () => {
            return {
                useDarkMode: vi.fn(() => [true, mockToggle]), // Устанавливаем isDark = true
            };
        });

        const { default: Navbar } = await import('./Navbar');

        render(
            <BrowserRouter>
                <Navbar />
            </BrowserRouter>
        );

        // Проверяем, что отображается иконка солнца (светлая тема)
        expect(screen.getByTestId('sunIcon')).toBeInTheDocument();
    });

    it('renderiza correctamente', async () => {
        vi.doMock('../hooks/useDarkMode', () => {
            return {
                useDarkMode: vi.fn(() => [false, mockToggle]),
            };
        });
        const { default: Navbar } = await import('./Navbar');

        render(
            <BrowserRouter>
                <Navbar />
            </BrowserRouter>
        );
        expect(screen.getByText(/Musk/)).toBeInTheDocument();
        expect(screen.getByText(/News/)).toBeInTheDocument();
    });

    it('activa el cambio de tema al hacer clic', async () => {
        vi.doMock('../hooks/useDarkMode', () => {
            return {
                useDarkMode: vi.fn(() => [false, mockToggle]),
            };
        });
        const { default: Navbar } = await import('./Navbar');

        render(
            <BrowserRouter>
                <Navbar />
            </BrowserRouter>
        );
        const toggleBtn = screen.getByTitle('switch theme');
        expect(screen.getByTestId('moonIcon')).toBeInTheDocument();
        fireEvent.click(toggleBtn);

        expect(mockToggle).toHaveBeenCalledTimes(1);
    });

    it('debe navegar a la categoría "General" al hacer clic en el enlace', async () => {
        vi.doMock('../hooks/useDarkMode', () => {
            return {
                useDarkMode: vi.fn(() => [false, mockToggle]),
            };
        });
        const { default: Navbar } = await import('./Navbar');

        render(
            <BrowserRouter>
                <Navbar />
            </BrowserRouter>
        );

        // Buscar el enlace "General"
        const generalLink = screen.getByText('General');

        // Simular clic
        fireEvent.click(generalLink);

        // Verificar que se haya llamado a navigate con la ruta correcta
        expect(mockNavigate).toHaveBeenCalledWith('/category/general');
    });
});
