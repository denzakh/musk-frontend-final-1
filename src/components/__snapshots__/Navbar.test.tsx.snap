// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Navbar > snapshot 1`] = `
<DocumentFragment>
  <nav
    class="relative bg-white dark:bg-gray-800 shadow p-4 flex justify-between items-center gap-4"
  >
    <div
      class="flex items-center gap-3 md:gap-4"
    >
      <a
        data-discover="true"
        href="/"
      >
        <div
          class="flex border-blue-500 border-2 font-semibold dark:border-blue-300"
        >
          <div
            class="bg-blue-500 text-white dark:text-gray-800 dark:bg-blue-300 px-2 py-1"
          >
            Musk
          </div>
          <div
            class="text-blue-500 dark:text-blue-200 bg-white dark:bg-gray-800 px-2 py-1"
          >
            News
          </div>
        </div>
      </a>
      <div
        class="shrink-0"
      >
        <button
          class="p-0 rounded hover:bg-gray-200 dark:hover:bg-gray-700 cursor-pointer flex items-center"
          title="switch theme"
        >
          <img
            alt=""
            class=""
            data-testid="moonIcon"
            height="30"
            src="mocked-moon.svg"
            width="30"
          />
        </button>
      </div>
    </div>
    <div
      class="hidden md:block"
    >
      <div
        class="bg-white dark:bg-gray-800 flex gap-8 md:gap-4 text-2xl md:text-base absolute md:relative flex-col md:flex-row top-18 md:top-0 right-1 md-right-0 z-10 p-8 md:p-0 shadow-xl md:shadow-none rounded-xl"
      >
        <a
          class="text-gray-600 dark:text-gray-300"
          data-testid="link-/category/general"
          href="/category/general"
        >
          General
        </a>
        <a
          class="text-gray-600 dark:text-gray-300"
          data-testid="link-/category/sports"
          href="/category/sports"
        >
          Sports
        </a>
        <a
          class="text-gray-600 dark:text-gray-300"
          data-testid="link-/category/technology"
          href="/category/technology"
        >
          Technology
        </a>
        <a
          class="text-gray-600 dark:text-gray-300"
          data-testid="link-/category/business"
          href="/category/business"
        >
          Business
        </a>
        <a
          class="text-gray-600 dark:text-gray-300"
          data-testid="link-/category/health"
          href="/category/health"
        >
          Health
        </a>
        <a
          class="text-gray-600 dark:text-gray-300"
          data-testid="link-/category/entertainment"
          href="/category/entertainment"
        >
          Entertainment
        </a>
      </div>
    </div>
    <div
      class="ml-[auto] md:ml-0 flex items-center gap-3 md:gap-4"
    >
      <a
        class="text-gray-600 dark:text-gray-300"
        data-testid="link-/filters"
        href="/filters"
      >
        <div
          class="flex items-center gap-2"
        >
          <img
            alt=""
            class=""
            height="30"
            src="mocked-search.svg"
            width="30"
          />
          <span
            class="hidden lg:inline"
          >
            Search
          </span>
        </div>
      </a>
      <a
        class="text-gray-600 dark:text-gray-300"
        data-testid="link-/favorites"
        href="/favorites"
      >
        <div
          class="flex items-center gap-2"
        >
          <img
            alt=""
            class=""
            height="30"
            src="mocked-favorite.svg"
            width="30"
          />
          <span
            class="hidden lg:inline"
          >
            Favorites
          </span>
        </div>
      </a>
    </div>
    <div
      class="flex items-center md:hidden"
    >
      <button
        class=""
      >
        <img
          alt=""
          class=""
          height="30"
          src="mocked-burger.svg"
          width="30"
        />
      </button>
    </div>
  </nav>
</DocumentFragment>
`;
